iceoryx2-ffi/cxx/tests/src/log_tests.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "iox/string.hpp"
+#include "iox/vector.hpp"
+#include "iox2/log.hpp"
+#include "iox2/log_level.hpp"
+
+#include "test.hpp"
+
+namespace {
+using namespace iox2;
+constexpr uint64_t TEST_LOGGER_CAPACITY = 10;
+constexpr uint64_t STRING_CAPACITY = 64;
+class Entry {
+  private:
+    LogLevel m_log_level;
+    iox::string<STRING_CAPACITY> m_origin;
+    iox::string<STRING_CAPACITY> m_message;
+
+  public:
+    Entry(LogLevel log_level, const char* origin, const char* message)
+        : m_log_level { log_level }
+        , m_origin { iox::TruncateToCapacity, origin }
+        , m_message { iox::TruncateToCapacity, message } {
+    }
+
+    auto is_equal(LogLevel log_level, const char* origin, const char* message) -> bool {
+        return m_log_level == log_level && m_origin == iox::string<STRING_CAPACITY>(iox::TruncateToCapacity, origin)
+               && m_message == iox::string<STRING_CAPACITY>(iox::TruncateToCapacity, message);
+    }
+};
+
+
+class TestLogger : public Log {
+  public:
+    static auto set_global_logger() {
+        auto& instance = get_instance();
+        set_logger(instance);
+    }
+
+    static auto get_instance() -> TestLogger& {
+        static TestLogger INSTANCE;
+        return INSTANCE;
+    }
+
+    void log(LogLevel log_level, const char* origin, const char* message) override {
+        if (m_log_buffer.size() < TEST_LOGGER_CAPACITY) {
+            m_log_buffer.emplace_back(log_level, origin, message);
+        }
+    }
+
+    auto get_log_buffer() -> iox::vector<Entry, TEST_LOGGER_CAPACITY> {
+        auto buffer = m_log_buffer;
+        m_log_buffer.clear();
+        return buffer;
+    }
+
+  private:
+    iox::vector<Entry, TEST_LOGGER_CAPACITY> m_log_buffer;
+};
+
+TEST(Log, custom_logger_works) {
+    TestLogger::set_global_logger();
+
+    log(LogLevel::Trace, "hello", "world");
+    log(LogLevel::Debug, "goodbye", "hypnotoad");
+    log(LogLevel::Info, "Who is looking for freedom?", "The Hoff!");
+    log(LogLevel::Warn, "Blümchen", "Bassface");
+    log(LogLevel::Error, "Blümchen should record a single with", "The almighty Hypnotoad");
+    log(LogLevel::Fatal, "It is the end", "my beloved toad.");
```

Comment:
Wouldn't it make sense to just use the iceoryx1 logging mechanism and forward the messages to the Rust logger?

We would basically get lazy evaluation for free and also logging complex messages without having to stringify it first.

Reply:
We have to talk about this. I lean to prefer that we have one single mechanism and that in the future - iceoryx2_hoofs would then also use it.

But the lazy evaluation and the existing infrastructure is a strong argument to not dismiss what we already have.

Maybe we can merge both of them together ...